File: src/main/java/org/oewntk/model/Synset.java
/*
 * Copyright (c) 2021-2021. Bernard Bou.
 */

package org.oewntk.model;

import java.io.Serializable;
import java.util.*;

/**
 * Synset
 */
public class Synset implements Comparable<Synset>, Serializable
{
	/**
	 * Synset id
	 */
	private final String synsetId;

	/**
	 * Synset type ss_type {n, v, a, r, s}
	 */
	private final char type;

	/**
	 * Lemma members
	 */
	private final String[] members;

	/**
	 * Definitions
	 */
	private final String[] definitions;

	/**
	 * Examples
	 */
	private final String[] examples;

	/**
	 * Wiki data
	 */
	private final String wikidata;

	/**
	 * Synset relations
	 */
	private Map<String, Set<String>> relations;

	/**
	 * Source file
	 */
	private final String domain;

	/**
	 * Constructor
	 *
	 * @param synsetId    synset id
	 * @param type        type: {n,v,a,r,s}
	 * @param domain      source file
	 * @param members     synset members
	 * @param definitions definitions
	 * @param examples    examples
	 * @param wikidata    wiki data
	 * @param relations   synset relations
	 */
	public Synset(final String synsetId, final char type, final String domain, final String[] members, final String[] definitions, final String[] examples, final String wikidata, final Map<String, Set<String>> relations)
	{
		this.synsetId = synsetId;
		this.type = type;
		this.domain = domain;
		this.members = members;
		this.definitions = definitions;
		this.examples = examples;
		this.wikidata = wikidata;
		this.relations = relations;
	}

	/**
	 * Get synset id
	 *
	 * @return synset id
	 */
	public String getSynsetId()
	{
		return synsetId;
	}

	/**
	 * Get type
	 *
	 * @return type
	 */
	public char getType()
	{
		return type;
	}

	/**
	 * Get part-of-speech
	 *
	 * @return part-of-speech
	 */
	public char getPartOfSpeech()
	{
		if (type == 's')
		{
			return 'a';
		}
		return type;
	}

	/**
	 * Get member lemmas
	 *
	 * @return member lemmas
	 */
	public String[] getMembers()
	{
		return members;
	}

	/**
	 * Get definitions
	 *
	 * @return definitions
	 */
	public String[] getDefinitions()
	{
		return definitions;
	}

	/**
	 * Get first definition
	 *
	 * @return first definition
	 */
	public String getDefinition()
	{
		if (definitions != null && definitions.length > 0)
		{
			return definitions[0];
		}
		return null;
	}

	/**
	 * Get examples
	 *
	 * @return examples
	 */
	public String[] getExamples()
	{
		return examples;
	}

	/**
	 * Get synset relations
	 *
	 * @return synset relations
	 */
	public Map<String, Set<String>> getRelations()
	{
		return relations;
	}

	/**
	 * Get wiki data
	 *
	 * @return wiki data
	 */
	public String getWikidata()
	{
		return wikidata;
	}

	/**
	 * Get domain
	 *
	 * @return source file
	 */
	public String getDomain()
	{
		return domain;
	}

	/**
	 * Get lex file
	 *
	 * @return lex file
	 */
	public String getLexfile()
	{
		switch (getPartOfSpeech())
		{
			case 'n':
				return "noun." + domain;
			case 'v':
				return "verb." + domain;
			case 'a':
				return "adj." + domain;
			case 'r':
				return "adv." + domain;
		}
		return null;
	}

	// mutation

	/**
	 * Add inverse synset relations of this synset
	 *
	 * @param inverseType    inverse type
	 * @param targetSynsetId target synset id
	 */
	public void addInverseRelation(final String inverseType, final String targetSynsetId)
	{
		if (relations == null)
		{
			relations = new HashMap<>();
		}
		var rels = relations.computeIfAbsent(inverseType, (k) -> new LinkedHashSet<>());
		if (rels.contains(targetSynsetId))
		{
			throw new IllegalArgumentException(String.format("%s duplicate %s to %s", getSynsetId(), inverseType, targetSynsetId));
		}
		rels.add(targetSynsetId);
	}

	// computed

	/**
	 * Find senses of this synset
	 *
	 * @param lexesByLemma lexes
	 * @return senses of this synset
	 */
	public Sense[] findSenses(Map<String, List<Lex>> lexesByLemma)
	{
		String[] members = getMembers();
		Sense[] senses = new Sense[members.length];
		int i = 0;
		for (String member : members)
		{
			for (Lex lex : lexesByLemma.get(member))
			{
				if (lex.getPartOfSpeech() != getPartOfSpeech())
				{
					continue;
				}
				for (Sense sense : lex.getSenses())
				{
					if (sense.getSynsetId().equals(getSynsetId()))
					{
						senses[i] = sense;
						i++;
					}
				}
			}
		}
		assert i == members.length;
		return senses;
	}

	/**
	 * Find sense of lemma in this synset
	 *
	 * @param lemma        lemma
	 * @param lexesByLemma lexes
	 * @return sense of lemma in this synset, null if not found
	 */
	public Sense findSenseOf(String lemma, Map<String, List<Lex>> lexesByLemma)
	{
		var lexes = lexesByLemma.get(lemma);
		assert lexes != null : String.format("%s has no sense", lemma);
		for (Lex lex : lexes)
		{
			if (lex.getPartOfSpeech() != getPartOfSpeech())
			{
				continue;
			}
			for (Sense sense : lex.getSenses())
			{
				if (sense.getSynsetId().equals(getSynsetId()))
				{
					return sense;
				}
			}
		}
		// System.err.printf("lemma %s not found in synset %s%n", lemma, this);
		return null;
	}

	/**
	 * Find index of member is members
	 *
	 * @param lemma member lemma
	 * @return index of member is members, -1 if not found
	 */
	public int findIndexOfMember(final String lemma)
	{
		String[] members = getMembers();
		List<String> memberList = Arrays.asList(members);
		return memberList.indexOf(lemma);
	}

	// stringify

	@Override
	public String toString()
	{
		String membersStr = Formatter.join(getMembers(), ",");
		String relationsStr = Formatter.join(getRelations(), ",");
		return String.format("%s %c {%s} '%s' {%s}", getSynsetId(), getType(), membersStr, getDefinition(), relationsStr);
	}

	// identity

	@Override
	public boolean equals(final Object o)
	{
		if (this == o)
		{
			return true;
		}
		if (o == null || getClass() != o.getClass())
		{
			return false;
		}
		Synset synset = (Synset) o;
		return synsetId.equals(synset.synsetId);
	}

	@Override
	public int hashCode()
	{
		return Objects.hash(synsetId);
	}

	// ordering

	@Override
	public int compareTo(final Synset that)
	{
		return getSynsetId().compareTo(that.getSynsetId());
	}
}

File: src/main/java/org/oewntk/model/Synset.kt
/*
 * Copyright (c) 2021-2024. Bernard Bou.
 */
package org.oewntk.model

import org.oewntk.model.Formatter.joinToString
import java.io.Serializable
import java.util.*

/**
 * Synset
 *
 * @param synsetId        synset id
 * @param type            synset ss_type {'n', 'v', 'a', 'r', 's'}
 * @param domain          source file
 * @param members         synset lemma members
 * @param definitions     definitions
 * @param examples        examples
 * @param usages          usages
 * @param wikidata        wiki data
 * @param relations       synset relations mapped by type
 *
 * @property synsetId     synset id
 * @property type         synset ss_type {'n', 'v', 'a', 'r', 's'}
 * @property domain       source file
 * @property members      synset lemma members
 * @property definitions  definitions
 * @property examples     examples
 * @property usages       usages
 * @property wikidata     wiki data
 * @property relations    synset relations mapped by type
 *
 * @property partOfSpeech synset part-of-speech {'n', 'v', 'a', 'r'} with ss_type 's' (satellite adj) mapped to 'a'
 * @property definition   first definition
 * @property lexfile      partOfSpeech.domain
 */
@kotlinx.serialization.Serializable
data class Synset(

    // key
    val synsetId: SynsetId,

    // value
    val type: SynsetType,
    val domain: Domain,
    val members: Set<Lemma>,
    val definitions: List<String>,
    val examples: List<Pair<String, String?>>? = null,
    val usages: List<String>? = null,
    var relations: Map<Relation, Set<SynsetId>>? = null,
    val ili: String? = null,
    val wikidata: List<String>? = null,

    ) : Comparable<Synset>, Serializable {

    var source: String? = null

    // computed properties (key, value)
    val key: SenseKey
        get() = synsetId
    val value: Array<Any?>
        get() = arrayOf(type, domain, members, definitions, examples, usages, relations, ili, wikidata, source)

    // computed properties
    val partOfSpeech: PartOfSpeech
        get() = type.toPartOfSpeech()
    val partOfSpeechName: String
        get() = partOfSpeech.fullName
    val definition: String?
        get() = if (definitions.isNotEmpty()) definitions[0] else null
    val lexfile: String = "$partOfSpeechName.$domain"
    val flatRelations: List<Pair<Relation, SynsetId>>?
        get() = relations?.flatMap { (key, values) -> values.map { key to it } }

    // identity

    override fun equals(other: Any?): Boolean {
        // throw UnsupportedOperationException("$this / $other")
        if (this === other) return true
        return if (other is Synset) {
            key == other.key
                    && value.contentEquals(other.value)
        } else false
    }

    override fun hashCode(): Int {
        // throw UnsupportedOperationException("$this")
        return Objects.hash(key, *value)
    }

    // ordering

    override fun compareTo(other: Synset): Int {
        return synsetId.compareTo(other.synsetId)
    }

    // mutation

    /**
     * Add inverse synset relations of this synset
     *
     * @param inverseType    inverse type
     * @param targetSynsetId target synset id
     */
    fun addInverseRelation(inverseType: Relation, targetSynsetId: SynsetId) {
        val mutableRelations = if (relations == null) HashMap() else relations!!.toMutableMap()
        relations = mutableRelations
        val inverseRelations =
            mutableRelations.computeIfPresent(inverseType) { _: Relation, v: Set<SynsetId> -> v.toMutableSet() }
                ?: mutableRelations.computeIfAbsent(inverseType) { LinkedHashSet() }

        require(!inverseRelations.contains(targetSynsetId)) { "Inverse relation $inverseType from $synsetId to $targetSynsetId was already there." }
        (inverseRelations as MutableSet<SynsetId>).add(targetSynsetId)
    }

    // computed

    /**
     * Find senses of this synset
     *
     * @param lemma2Lexes lemma to lexes mapper
     * @param senseKey2Sense senseKey to sense mapper
     * @return senses of this synset
     */
    fun findSenses(
        lemma2Lexes: (lemma: Lemma) -> Collection<Lex>,
        senseKey2Sense: (senseKey: SenseKey) -> Sense,
    ): List<Sense> {

        return members
            .asSequence()
            .map { lemma2Lexes(it) }
            .flatMap { it.asSequence() }
            .filter { it.partOfSpeech == this.partOfSpeech }
            .flatMap { it.senseKeys }
            .map { sk -> senseKey2Sense(sk) }
            .filter { s -> s.synsetId == this.synsetId }
            .toList()
    }

    /**
     * Find sense of lemma in this synset
     *
     * @param lemma lemma
     * @param lexResolver lemma to lexes resolver
     * @param senseKeyResolver senseKey to sense resolver
     * @return sense of lemma in this synset, null if not found
     */
    @Throws(IllegalStateException::class)
    fun findSenseOf(
        lemma: Lemma,
        lexResolver: (Lemma) -> Collection<Lex>,
        senseKeyResolver: (SenseKey) -> Sense,
    ): Sense? {
        val lexes: Collection<Lex> = lexResolver(lemma)
        return lexes
            .asSequence()
            .filter { it.partOfSpeech == partOfSpeech }
            .flatMap { it.senseKeys }
            .map { sk -> senseKeyResolver(sk) }
            .firstOrNull { s -> s.synsetId == this.synsetId }
    }

    /**
     * Find sense of lemma in this synset
     *
     * @param lemma lemma
     * @param lexResolver lemma to lexes resolver
     * @param senseKeyResolver senseKey to sense resolver
     * @return sense of lemma in this synset, null if not found
     */
    @Throws(IllegalStateException::class)
    fun resolveSenseOf(
        lemma: Lemma,
        lexResolver: (lemma: Lemma) -> Collection<Lex>,
        senseKeyResolver: (senseKey: SenseKey) -> Sense,
    ): Sense {
        return findSenseOf(lemma, lexResolver, senseKeyResolver) ?: throw IllegalStateException("Lemma '$lemma' not found in synset $this")
    }

    /**
     * Find 0-based index of member is members
     *
     * @param lemma member lemma
     * @return index of member is members, -1 if not found
     */
    fun findIndexOfMember(lemma: Lemma): Int = members.toList().indexOf(lemma)

    // stringify

    override fun toString(): String {
        val membersStr = members.joinToString(",")
        val relationsStr = relations?.get("hypernym") ?: ""
        return "$synsetId ${type.value} {$membersStr} '$definition' ^$relationsStr"
    }

    fun toLongString(): String {
        val membersStr = members.joinToString(",")
        val relationsStr = relations?.joinToString(",") ?: ""
        return "$synsetId ${type.value} {$membersStr} '$definition' {$relationsStr}"
    }

    companion object {

        val VALID_SYNSET_RELATIONS = arrayOf(

            "hypernym",  // "hyponym",
            "instance_hypernym",  // "instance_hyponym",
            "mero_part",  // "holo_part",
            "mero_member",  // "holo_member",
            "mero_substance",  // "holo_substance",
            "causes",  // "is_caused_by",
            "entails",  // "is_entailed_by",
            "exemplifies",  // "is_exemplified_by",
            "domain_topic",  // "has_domain_topic"
            "domain_region",  // "has_domain_region"
            "attribute",
            "similar",
            "also",
        )

        val SYNSET_RELATIONS = arrayOf(

            "hypernym", "hyponym",
            "instance_hypernym", "instance_hyponym",
            "mero_part", "holo_part",
            "mero_member", "holo_member",
            "mero_substance", "holo_substance",
            "causes", "is_caused_by",
            "entails", "is_entailed_by",
            "exemplifies", "is_exemplified_by",
            "domain_topic", "has_domain_topic",
            "domain_region", "has_domain_region",
            "attribute",
            "similar",
            "also",
        )

        val INVERSE_SYNSET_RELATIONS = mapOf(
            "hypernym" to "hyponym",
            "instance_hypernym" to "instance_hyponym",
            "mero_part" to "holo_part",
            "mero_member" to "holo_member",
            "mero_substance" to "holo_substance",
            "causes" to "is_caused_by",
            "entails" to "is_entailed_by",
            "exemplifies" to "is_exemplified_by",
            "domain_topic" to "has_domain_topic",
            "domain_region" to "has_domain_region",
        )

        /*
        ignored

        agent|
        // also|
        // attribute|
        be_in_state|
        // causes|
        classified_by|
        classifies|
        co_agent_instrument|
        co_agent_patient|
        co_agent_result|
        co_instrument_agent|
        co_instrument_patient|
        co_instrument_result|
        co_patient_agent|
        co_patient_instrument|
        co_result_agent|
        co_result_instrument|
        co_role|
        direction|
        // domain_region|
        // domain_topic|
        // exemplifies|
        // entails|
        eq_synonym|
        // has_domain_region|
        // has_domain_topic|
        // is_exemplified_by|
        holo_location|
        // holo_member|
        // holo_part|
        holo_portion|
        // holo_substance|
        holonym|
        // hypernym|
        // hyponym|
        in_manner|
        // instance_hypernym|
        // instance_hyponym|
        instrument|
        involved|
        involved_agent|
        involved_direction|
        involved_instrument|
        involved_location|
        involved_patient|
        involved_result|
        involved_source_direction|
        involved_target_direction|
        // is_caused_by|
        // is_entailed_by|
        location|
        manner_of|
        mero_location|
        // mero_member|
        // mero_part|
        mero_portion|
        // mero_substance|
        meronym|
        // similar|
        // other|
        patient|
        restricted_by|
        restricts|
        result|
        role|
        source_direction|
        state_of|
        target_direction|
        subevent|
        is_subevent_of|
        // antonym|
        feminine|
        has_feminine|
        masculine|
        has_masculine|
        young|
        has_young|
        diminutive|
        has_diminutive|
        augmentative|
        has_augmentative|
        anto_gradable|
        anto_simple|
        anto_converse|
        ir_synonym
       */

    }
}
